### DISCUSSION TITLE
Dynamically changing getlisteners()

### DISCUSSION THREAD

**Discussion by 94Joe** [livewire/livewire#3655] [1. Help] [answered]
I'm using [beyondcode/laravel-websockets](https://github.com/beyondcode/laravel-websockets) with [laravel/echo](https://github.com/laravel/echo) on localhost. Public channels work fine, I also tested private channels by giving productId manually and it works. 

The problem I'm facing is how to make livewire listen to the private channel event once the product is created? because right now I see `500 Internal Server Error` and according to `laravel.log` it says `Too few arguments to function...` which is normal because my route is expecting a productId which is not set yet. It looks like `getListeners()` is only called on inital `render/mount`. Is there a way to update `getListeners()` ? 

Modules/Product/Routes/**channels.php**
````
Broadcast::channel('products.{productId}', function ($user, $productId) {
    return $user->id === Product::findOrFail($productId)->user_id;
});
````

Modules/Product/Http/Livewire/**Products.php**
```
class Products extends Component
{
    public $productId;

    public function getListeners()
    {
        return ["startListening" => 'triggered', "echo-private:products.{$this->productId},\\Modules\\Product\\Events\\ProductCreated" => 'test'];
    }

    public function triggered($productId)
    {
        $this->productId= $productId;
    }

    public function test()
    {
        dd('ok');
    }

    public function render()
    {
        return view('product::livewire.index');
    }
}
```
PS:  I'm sending the productId from another component once the product is created using `$this->emit('startListening',  $product->id)` .

**Comment by joshhanley** [comment, marked as answer]
@94Joe currently there is no way to change the listeners once they are running. 

To work around it, you would need to generate a listener in javascript instead that just calls Livewire `$refresh` or something like that.

Hope this helps!
